### PR TITLE
feat(rust,python): cast each parquet file to delta schema

### PR DESCRIPTION
**Pull request by HawaiianSpork**
# Description

By casting the read record batch to the delta schema datafusion can read tables where the underlying parquet files can be cast to the desired schema. Fixes:

- Errors querying data where some of the parquet files may not have columns that were added later because of schema migration.  This includes nested columns for structs that are in Maps, Lists, or children of other structs
- maps and lists written with different different element names
- timestamps of different units.
- Any other cast supported by arrow-cast.

This can be done now since data-fusion exposes a SchemaAdapter which can be overwritten.

We should note that this makes all times being read by delta-rs as having microsecond precision to match the Delta protocol.

# Related Issue(s)
- This makes solving #2478 and #2341 just a matter of adding code to delta-rs cast.

### CODE REVIEW

File: crates/core/src/delta_datafusion/mod.rs [hunk 1]
```diff
@@ -618,35 +615,30 @@ impl<'a> DeltaScanBuilder<'a> {
...
-            None
+        if let Some(predicate) = logical_filter {
+            if config.enable_parquet_pushdown {
+                exec_plan_builder = exec_plan_builder.with_predicate(predicate);
+            }
         };
 
-        let scan = ParquetFormat::new()
-            .create_physical_plan(
-                self.state,
```
Comment (HawaiianSpork) [inline crates/core/src/delta_datafusion/mod.rs:632]: State was previously passed in but was not used by the physical plan.

File: crates/core/src/delta_datafusion/mod.rs [hunk 2]
```diff
@@ -618,35 +615,30 @@ impl<'a> DeltaScanBuilder<'a> {
...
+            object_store_url: self.log_store.object_store_url(),
+            file_schema,
+            file_groups: file_groups.into_values().collect(),
+            statistics: stats,
+            projection: self.projection.cloned(),
+            limit: self.limit,
+            table_partition_cols,
+            output_ordering: vec![],
+        })
+        .with_schema_adapter_factory(Arc::new(DeltaSchemaAdapterFactory {}));
```
Comment (ion-elgreco) [inline crates/core/src/delta_datafusion/mod.rs:628]: Why is the correct schema not passed in the factory?

Comment (HawaiianSpork) [reply crates/core/src/delta_datafusion/mod.rs:624]: oops, fixed

**Comment by ion-elgreco** [comment]
@HawaiianSpork can you add a  test where  we have  a delta table that contains parquets with nanosecond timestamps in the files. Maybe just create a parquet table and then use convert to delta?

**Comment by HawaiianSpork** [comment]
> @HawaiianSpork can you add a test where we have a delta table that contains parquets with nanosecond timestamps in the files. Maybe just create a parquet table and then use convert to delta?

@ion-elgreco, I'd be happy to add more tests but want to make sure I create the correct ones. `../test/tests/data/table_with_edge_timestamps` data has parquet files with nanosecond timestamp precision, you can see how this change leads datafusion only seeing microsecond precision.  Do you think this is fine?

FYI, @wjones127 and @roeap for making the original commit that read the schema from the parquet files: #1266.

**Comment by rtyler** [comment]
This looks promising, but I would like to update the title if you don't mind for the changelog in the future. Schema evolution is typically understood in the Delta context as changes to the Delta schema (i.e. a transaction commit occurs). 

I am understanding this correctly it's more about schema adaptation on reads

**Review by ion-elgreco — APPROVED** [review]

**Comment by ion-elgreco** [comment]
@HawaiianSpork can you rebase so that we can merge?

**Comment by HawaiianSpork** [comment]
@ion-elgreco I have merged with master as requested.

**Pull request merged**